Adding new color mode to mesh plot to cycle colors like curves

markcmiller86 opened the pull request:
### Description
Resolves #3465 
Adds new color options for both mesh and opaque colors of mesh plot to change the color upon each new plot instantiation.

I searched for how to define colors that are maximally distinct from each other since the intended goal is to have multiple mesh plots (or sub-parts thereof), that can be distinguished. Best I could find was a static list of 20 maximally distinct colors. I think having more than 20 mesh plots simultaneously in the same window is a low likelihood so I think 20 colors is sufficient.

I don't wanna call the control *Cycle* (as in Curve plots) because I think its to easy to confuse that with referring to cycles in a time series. I called it *Random*. For the user, the color is essentially random. It doesn't *cycle* back to a previously seen color until it reaches the maximum of 20. Another name for this color mode could be *Iterative* or *Iterate*.

### Type of change
- [ ] Bug fix
- [x] New feature
- [ ] New Documentation
- [ ] Other (please describe below)

### How Has This Been Tested?

Just starting the PR. Won't be done for another day or two.

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the release notes
- [ ] I have made corresponding changes to the documentation
- [ ] I have added debugging support to my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added any new baselines to the repo
- [x] I have assigned reviewers (see [VisIt's PR procedures](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers) for more information).

markcmiller86 committed:
snapshot

biagas commented on src/plots/Mesh/QvisMeshPlotWindow.C:
```diff
@@ -205,9 +205,12 @@ QvisMeshPlotWindow::CreateWindowContents()
     meshColorButtons->addButton(rb, 0);
     colorLayout->addWidget(rb, 0, 1);
+    rb = new QRadioButton(tr("Random"), central);
+    meshColorButtons->addButton(rb, 2);
```
Just a nit, but if 'Random' comes before 'Custom', shouldn't its' id be smaller (eg 1), and Custom's id be 2?

markcmiller86 replied:
Well, its third in the enum but second in the GUI widget set (foreground, random, custom). The reason being is that custom *must* go at the end because there is the color selector box that follows it. 
![Untitled](https://user-images.githubusercontent.com/5720676/58644950-37e6a800-82b7-11e9-972f-76a62720c255.png)

biagas replied:
Yeah, I understood that about the gui. Also, they currently need to be the id you've chosen because of the order they appear in the atts enum. You can ignore me.

biagas commented on src/plots/Mesh/MeshAttributes.h:
```diff
@@ -71,12 +71,14 @@ class MeshAttributes : public AttributeSubject
         Foreground,
-        MeshCustom
+        MeshCustom,
+        MeshRandom
```
I am fine with 'Random'

markcmiller86 committed:
snapshot

markcmiller86 committed:
Merge branch 'develop' into feature/markcmiller86/cycle_mesh_colors_29May19

markcmiller86 committed:
Finish up changes

markcmiller86 commented:
I've taken this out of *draft* mode. I think its ready.

I've yet to add tests and will do so before committing.

biagas reviewed (dismissed):
+1

biagas commented on src/plots/Mesh/MeshAttributes.code:
```diff
@@ -189,3 +189,103 @@ MeshAttributes::ProcessOldVersions(DataNode *parentNode,
+    {145,  30, 180, 255}, { 70, 240, 240, 255}, {240,  50, 230, 255}, {210, 245,  60, 255}, {250, 190, 190, 255},
+    {  0, 128, 128, 255}, {230, 190, 255, 255}, {170, 110,  40, 255}, {255, 250, 200, 255}, {128,   0,   0, 255},
+    {170, 255, 195, 255}, {128, 128,   0, 255}, {255, 215, 180, 255}, {  0,   0, 128, 255}, {128, 128, 128, 255},
+    {255, 255, 255, 255}, {  0,   0,   0, 255}};
```
Would it make sense to create a Discrete color table with these values and use it instead.  It might make a good alternative for Contour plots and others that make use of Discrete color tables.

markcmiller86 replied:
Thats a good idea.

Would I then change to this color map when *random* mode is selected and then change back to the old colormap after *random* is switched to something else?

biagas commented:
On Mon, 2019-06-10 at 12:01 -0700, Mark C. Miller wrote:
> @markcmiller86 commented on this pull request.
> 
> 
> ______________________________________________________________________
> 
> In src/plots/Mesh/MeshAttributes.code:
> 
> > +//  Arguments:
> +//    atts      The attribute subject to initialize.
> +//    md        The metadata used to initialize.
> +//    atts      The variable name used to initialize.
> +//
> +//  Programmer: miller86 -- generated by xml2info
> +//  Creation:   Mark C. Miller, Thu May 30 22:41:29 PDT 2019
> +// ****************************************************************************
> +static int colorIndex = 0;
> +// Grabbed 29May19 from https://sashat.me/2017/01/11/list-of-20-simple-distinct-colors/
> +static unsigned char SashaTrubetskoy22SimpleDistinctColors[][4] = {
> +    {230,  25,  75, 255}, { 60, 180,  75, 255}, {255, 225,  25, 255}, {  0, 130, 200, 255}, {245, 130,  48, 255},
> +    {145,  30, 180, 255}, { 70, 240, 240, 255}, {240,  50, 230, 255}, {210, 245,  60, 255}, {250, 190, 190, 255},
> +    {  0, 128, 128, 255}, {230, 190, 255, 255}, {170, 110,  40, 255}, {255, 250, 200, 255}, {128,   0,   0, 255},
> +    {170, 255, 195, 255}, {128, 128,   0, 255}, {255, 215, 180, 255}, {  0,   0, 128, 255}, {128, 128, 128, 255},
> +    {255, 255, 255, 255}, {  0,   0,   0, 255}};
> 
> Thats a good idea.
> 
> Would I then change to this color map when random mode is selected and
> then change back to the old colormap after random is switched to
> something else?
> 
I don't think you need to change color map at all. You can just
reference it when setting the Random color, the way that Curve plot
references the 'Default discrete' color table.  See
CurveViewerEnginePluginInfo::SetColor.
You could just reference whatever name you give this color table.

Built in color tables are created in avtColorTables.C, others live as ct
files in resources/colortables.  All are loaded on startup, though if
you want to reference it by name explicitly (rather than via
DefaultDiscreteColorTable). you may want to ensure it exists first, in
case user removed it via saved configuration settings.

Kathleen

> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

markcmiller86 commented:
Ah, Ok, I was thinking that Mesh plot had a color map setting. It doesn't. So, the idea you are suggesting about using the colors here as a color map object is really only about implementation. It doesn't have anything to do with the Mesh plot interface.

markcmiller86 commented:
@biagas ... in `avtColorTables.C`, it appears that control points are specified with RGBA (e.g. includes *alpha*) but then the alpha is[ actually ignored ](https://github.com/visit-dav/visit/blob/12853788e81721b6e50bcb4bd4cf78ffaef89999/src/avt/Pipeline/Data/avtColorTables.C#L576)in this data when `ColorControlPointList` is created. 

Now, I noticed that even `hot` color table has non unity alpha in some of its control points. And, I am wondering if its ok for my new color table there to have unity for all control point alphas?

biagas commented:
Yes, it is fine to have all unity for alphas.

Kathleen

On Mon, 2019-06-10 at 16:33 -0700, Mark C. Miller wrote:
> @biagas ... in avtColorTables.C, it appears that control points are
> specified with RGBA (e.g. includes alpha) but then the alpha is
> actually ignored in this data when ColorControlPointList is created.
> 
> Now, I noticed that even hot color table has non unity alpha in some
> of its control points. And, I am wondering if its ok for my new color
> table there to have unity for all control point alphas?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

markcmiller86 committed:
make new distinct color table and use that

markcmiller86 commented:
@biagas I've added a new `distinct` color table and adjusted code as you suggested. That said, should I only use default, discrete setting and let user set that or should I explicitly attempt to use distinct and if that fails, fall back to whatever default discrete ct is? I've coded the latter in this PR

biagas reviewed (changes requested):
I downloaded and compiled to take a look at the color table. That part looks good.
I had a problem using the 'Random' setting.  If I create a Mesh plot and select 'Random', I always get black.  The only way I really get the Random coloring is if 'Random' is the default. (eg Change selection to 'Random', click 'make default'.)  I'm not sure off the top of my head if there is another way to implement.  If not, at the very least, the release notes and Mesh plot documentation need to state that this new option requires user to make the option the default.

markcmiller86 commented:
Yeah, I noticed that too. I was thinking that was considered *correct* behavior though. But, now that I see you write the description of behavior...I think we may want the selection of *random* to actually result in a change of the current color to the current *random* color. Let me work on that.

biagas commented:
I also thought it strange that even if I changed to 'Random' and 'Make Default' before I clicked Draw, it still didn't have any effect.

markcmiller86 commented:
I guess a key question I've yet to fully answer for myself is *when* is the choice of the random color made?
* When*ever* **Random** color option is selected? If so, wouldn't that mean a different random color is selected each time that button is hit, even for the same mesh plot?
* The *first* time **Random** color option is selected for a plot and forever after that particular plot instance has that random color?
* Only when a plot is *created*, which I think requires default plot attributes to have **Random** enabled (which is what I originally implemented but now don't like).

If I create several mesh plots with **Random** enabled and each gets its random color, then delete them all, then create a possibly different set of mesh plots, should I expect the new mesh plots to get the same colors in the same sequence?

Finally, I neglected one key issue and that is to ensure the **Random** color is never one that too closely matches the background color.

markcmiller86 committed:
snapshot

markcmiller86 commented:
Does anyone know if VisIt has any internal functionality for manipulating colors in HSL space instead of RGB? How about VTK?

biagas commented:
vtkLookupTable has HSV range settings. 
https://vtk.org/doc/nightly/html/classvtkLookupTable.html

VisIt sets HueRange in avtLookupTable, only to set up a default 'hot'
color table. I believe our color tables override this default.

Kathleen

On Tue, 2019-06-18 at 09:35 -0700, Mark C. Miller wrote:
> Does anyone know if VisIt has any internal functionality for
> manipulating colors in HSL space instead of RGB? How about VTK?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

markcmiller86 committed:
improving behavior

markcmiller86 commented:
Updated with several improvements to behavior.

- Fixed some issues in Curve plot's implementation of color cycling
  - possible UMR in ColorAttribute
  - non-sensicle mod operator with GetNumColors()
- Added `instanceIndex` to `avtPlot`. `index` alone was not sufficient because its value isn't defined until long after an `avtPlot` is created.
- Mesh plot's random color logic uses `instanceIndex` similarly to Curve plot's `colorIndex` static variable. It would be better to use `avtPlot::index` but because that value is `-1` in the early life of a plot, it can lead to many Mesh plot objects selecting the same color. If we could somehow delay the selection of the randomly chosen color until after the `avtPlot::index` has been set, that would help
- Mesh plot color selection logic attempts to select mesh color that is sufficiently distant from foreground color and opaque color that is sufficientl distant from background color.
- I introduced color distance functions to avtColorTables.

biagas commented on src/avt/Pipeline/Data/avtColorTables.C:
```diff
@@ -1099,3 +1140,24 @@ avtColorTables::ImportColorTables()
+
+// ****************************************************************************
+// Method: avtColorTables::ImportColorTables
+//
```
Please make the method name in the comments match the method.

biagas commented on src/plots/Mesh/avtMeshPlot.C:
```diff
@@ -347,21 +349,71 @@ avtMeshPlot::SetAtts(const AttributeGroup *a)
+            colorIndex++;
+        }
+        while ((colorIndex - this->instanceIndex) < 50 &&
+               (ct->PerceptualColorDistance(rgb, fg) <= ct->JustNoticeableColorDistance()));
```
So this is cool. But I think you should compare against background color not foreground. I don't think we care  if the foreground color is randomly used, but we do care if the mesh disappears into the background.

markcmiller86 replied:
Oh, of course. Good catch!

biagas commented on src/plots/Mesh/avtMeshPlot.C:
```diff
@@ -347,21 +349,71 @@ avtMeshPlot::SetAtts(const AttributeGroup *a)
+            colorIndex++;
+        }
+        while ((colorIndex - this->instanceIndex) < 50 &&
+               (ct->PerceptualColorDistance(rgb, bg) <= ct->JustNoticeableColorDistance()));
```
While very cool, does it really matter if the 'opaque' surface randomly becomes the same color as the background? That is the default mode and perhaps a valid color even in random mode?

markcmiller86 replied:
Maybe it doesn't matter. I does matter if the opaque color randomly becomes the same color as the mesh color. However, the only way I have to try to avoid that was to start at a different index in the colortable, hence the `+1` in initializing `colorIndex`. I don't have a way to ask the mesh what is current color is.

biagas reviewed (changes requested):
I like your shift of colors so that we don't get washout with background. Can you add it for curves, too?
I set your new distinct color table as the active/default Discrete color table, then did a bunch of lineouts.
Once curve eventually picked up the background color and washed out.

markcmiller86 commented:
> I like your shift of colors so that we don't get washout with background. Can you add it for curves, too?
> I set your new distinct color table as the active/default Discrete color table, then did a bunch of lineouts.
> Once curve eventually picked up the background color and washed out.

I can...to do so means that the Curve plot has to know what the background color is. Is there an easy way to do that or do I need to emulate what Mesh plot does and hold a copy `bgColor` in it that is somehow kept up to date by its producers and consumers?

biagas commented:
Yes, I guess Curve plot will have do the same as Mesh. 
I had only suggested you modify Curve since you had already made 1
change.
On Wed, 2019-06-19 at 10:22 -0700, Mark C. Miller wrote:
>         I like your shift of colors so that we don't get washout with
>         background. Can you add it for curves, too?
>         I set your new distinct color table as the active/default
>         Discrete color table, then did a bunch of lineouts.
>         Once curve eventually picked up the background color and
>         washed out.
>         
> 
> I can...to do so means that the Curve plot has to know what the
> background color is. Is there an easy way to do that or do I need to
> emulate what Mesh plot does and hold a copy bgColor in it that is
> somehow kept up to date by its producers and consumers?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

markcmiller86 commented:
> Yes, I guess Curve plot will have do the same as Mesh.

I think its as simple is overriding the `SetBackgroundColor` method for `avtPlot` in the curve plot. Agreed?

markcmiller86 commented:
@biagas ... problem here...I've just about Curve plot coded and realized, in Mesh plot, this color logic is handled in the `avtMeshPlot` object which knows its background color whereas in Curve, it is handled in `CurveViewerPluginInfo::SetColor` which is called from `CurveViewerEnginePluginInfo::InitializePlotAtts()`. I like doing stuff in the PluginInfo object because its really early in creation of the plot object I think, right? But, to do the work of distinguishing colors, I also need to know about colors in use for things like background, maybe foreground, would be nice to know what colors are in use by other plot objects in the plot list. OTOH, if the color logic is handled *only* in `XXXViewerEnginePluginInfo` then that also means it can be manipulated only at plot creation and not changed once the plot is in the plot list, which is undesireable. So, for Curve plots, I think I need to move to be more like new mesh plot. Thoughts?

biagas commented:
Just remembered that setting curve colors is complicated by the fact
that global lineout options may also affect curve color.
Perhaps the suggested changes to curve plot are outside the scope of the
issue you are addressing. I think a new ticket should be created, it
will allow more time for proper consideration of how the global lineout
settings would affect the changes you suggest for curve plot mimicking
mesh plot in how it sets colors.

On Wed, 2019-06-19 at 19:16 -0700, Mark C. Miller wrote:
> @biagas ... problem here...I've just about Curve plot coded and
> realized, in Mesh plot, this color logic is handled in the avtMeshPlot
> object which knows its background color whereas in Curve, it is
> handled in CurveViewerPluginInfo::SetColor which is called from
> CurveViewerEnginePluginInfo::InitializePlotAtts(). I like doing stuff
> in the PluginInfo object because its really early in creation of the
> plot object I think, right? But, to do the work of distinguishing
> colors, I also need to know about colors in use for things like
> background, maybe foreground, would be nice to know what colors are in
> use by other plot objects in the plot list. OTOH, if the color logic
> is handled only in XXXViewerEnginePluginInfo then that also means it
> can be manipulated only at plot creation and not changed once the plot
> is in the plot list, which is undesireable. So, for Curve plots, I
> think I need to move to be more like new mesh plot. Thoughts?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

markcmiller86 committed:
move random color selection logic to avtColorTable class

markcmiller86 committed:
update to use random color selection method in avtColorTable

markcmiller86 committed:
update Curve plot to operate similarly to Mesh plot

markcmiller86 committed:
Merge branch 'develop' into feature/markcmiller86/cycle_mesh_colors_29May19

markcmiller86 commented:
@biagas ... I think I have working Curve capability now too. I've also improved overall implementation in a few ways. Please have a look when you get a chance.

markcmiller86 committed:
removing changes to internallauncher to sep. branch

biagas commented on src/plots/Mesh/MeshViewerEnginePluginInfo.C:
```diff
@@ -205,3 +209,100 @@ MeshViewerEnginePluginInfo::GetMenuName() const
+        colorIndex = (colorIndex + 1) % ct->GetNumColors();
+    }
+}
+
```
Is the code in the Plugin Info still needed? If so, should it use your new avtColorTable method?

markcmiller86 replied:
Looks like I failed to re-gen that. Sorry. Let me re-gen really quickly. No, it is not needed.

markcmiller86 replied:
Ok, I just pushed that fix. Sorry 'bout that.

markcmiller86 replied:
oops. no I didn't. I fixed the *same* problem left in Curve plot. Let me fix mesh plot too. Doh!

markcmiller86 replied:
Ok, pushed fixes for this too.

markcmiller86 committed:
regen due to change in xml file

markcmiller86 committed:
removing extraneous code from older approach

biagas commented on src/avt/Pipeline/Data/avtColorTables.C:
```diff
@@ -998,6 +1039,56 @@ avtColorTables::GetControlPointColor(const std::string &ctName, int i,
 
+
+
+
```
Sorry, small nit, an awful lot of whitespace here.

markcmiller86 committed:
remove extraneous whitespace

markcmiller86 commented:
Sorry, you will have to re-approve as I made the whitespace change you requested invalidating the approval you made.

Mark



"Kathleen Biagas" wrote:


@biagas approved this pull request.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/visit-dav/visit/pull/3534?email_source=notifications&email_token=ABLUUZDVAQKHKIFKGCGYBXDP3QLFNA5CNFSM4HQ656ZKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4HD2VQ#pullrequestreview-252591446>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABLUUZCHPCZY4FMRG6V6YY3P3QLFNANCNFSM4HQ656ZA>.

biagas commented:
there is no longer an option for me to review.

markcmiller86 commented:
> there is no longer an option for me to review.

Ok, I just manually requested a re-review

markcmiller86 committed:
adding closedb calls and starting new section for cycling colors

markcmiller86 committed:
finishing merge

markcmiller86 committed:
fix snprintf merge

markcmiller86 committed:
adding tests

markcmiller86 committed:
adding documentation

markcmiller86 commented:
@biagas...I think I've completed. Added to docs. Added tests. Carried functionality to curves as much as possible too.

markcmiller86 committed:
fixing typo in link/anchor

markcmiller86 committed:
Merge branch 'develop' into feature/markcmiller86/cycle_mesh_colors_29May19

biagas commented on src/plots/Curve/CurveAttributes.code:
```diff
@@ -253,11 +253,9 @@ Definition:
 void
 CurveViewerEnginePluginInfo::InitializeGlobalObjects()
 {
-    colorIndex = 0;
```
Since this code is no longer custom, it can be removed from the .code and .xml files.
Same with InitializePlotAtts.

biagas replied:
InitializePlotAtts still needs to be removed.

biagas commented on src/plots/Mesh/MeshAttributes.code:
```diff
@@ -189,3 +187,30 @@ MeshAttributes::ProcessOldVersions(DataNode *parentNode,
     }
 }
+
+Target: xml2info
```
This method contains no custom logic, so isn't needed.

markcmiller86 replied:
I'm confused. I didn't touch this method and don't understand the rationale for removing it if that is what you are requesting. It seems to me like it is custom to handle reading of mesh attributes that pre-date 2.0.0 and 3.0.0.

biagas replied:
I apologize if the positioning of my comment caused confusion. It's referring to 'InitializePlotAtts' which you added to the code file.

biagas replied:
InitializePlotAtts still needs to be removed from the .code file.

biagas commented on src/plots/Mesh/MeshAttributes.h:
```diff
@@ -37,12 +37,14 @@ class MeshAttributes : public AttributeSubject
         Foreground,
-        MeshCustom
+        MeshCustom,
+        MeshRandom
```
Have you checked whether old python scripts or saved configurations still work?  Just wanting to see if any coding is needed to process old settings.

markcmiller86 replied:
I haven't tested that. Good suggestion. Will try later this week. This case in particular, the added values to the enum should NOT be a problem except if trying to read *new* session files or scripts into an older VisIt. Are we required to support that too? Or just new VisIt reading old scripts/session files?

markcmiller86 replied:
@biagas can you point me in the right direction for testing this. Where do I start?

markcmiller86 replied:
Ok, in lieu of any further guidance, I ran all current session tests. They all pass. I think this is good to go now.

markcmiller86 committed:
removing code as per @biagas

markcmiller86 committed:
Merge branch 'feature/markcmiller86/cycle_mesh_colors_29May19' of github.com:visit-dav/visit into feature/markcmiller86/cycle_mesh_colors_29May19

markcmiller86 committed:
fix removal of InitializePlotAtts
